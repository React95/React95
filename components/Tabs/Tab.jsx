import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const NavItem = styled.li`
  list-style: none;
  padding: 3px 6px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff, -2px 1px 0 -1px #c3c7cb,
    -3px 1px 0px -1px #ffffff, -2px 0px 0px -1px #ffffff, 2px 1px 0 -1px #868a8e,
    2px 0px 0 -1px #000, 3px 1px 0 -1px #000;
  margin-right: 2px;
  margin-left: 2px;

  &:first-child {
    margin-left: 4px;
  }

  ${props => props.active
    && css`
      margin-top: -2px;
      margin-left: -2px;
      margin-bottom: 1px;

      &:first-child {
        margin-left: 2px;
      }

      box-shadow: inset 0px 1px 0px 0px #ffffff, -2px 1px 0 -1px #c3c7cb,
        -0.5px 3px 0 0.5px #c3c7cb, -2px 2px 0px 0px #fff,
        -2px 0px 0px -1px #ffffff, 1px 2px 0 0px #868a8e, 2px 0px 0 -1px #000,
        2px 2px 0 0px #000;

      & + li {
        margin-left: 0;
        box-shadow: inset 0px 1px 0px 0px #ffffff, 2px 1px 0 -1px #868a8e,
          2px 0px 0 -1px #000, 3px 1px 0 -1px #000;
      }
    `}
`;

class Tab extends Component {
  constructor(props) {
    super(props);
  }

  onClick = () => {
    const { onClick } = this.props;
    onClick(this);
  };

  render() {
    const { activeTab, title } = this.props;

    return (
      <NavItem active={activeTab === title} onClick={this.onClick}>
        {title}
      </NavItem>
    );
  }
}

Tab.propTypes = {
  activeTab: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Tab;
