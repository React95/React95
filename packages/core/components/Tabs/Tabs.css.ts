import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';
import { recipe } from '@vanilla-extract/recipes';
import { calc } from '@vanilla-extract/css-utils';

export const navbar = style({
  display: 'flex',
  padding: '0',
  margin: '0',
  border: 'none',
  zIndex: 1,
});

export const navContainer = style({
  padding: contract.space[12],
  boxShadow: contract.shadows.out,
  backgroundColor: contract.colors.material,
});

export const tab = recipe({
  base: {
    listStyle: 'none',
    userSelect: 'none',
    paddingBlock: contract.space[3],
    paddingInline: contract.space[6],
    backgroundColor: contract.colors.material,
    marginRight: contract.space[2],
    marginLeft: contract.space[2],
    boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
      -2px 1px 0 -1px ${contract.colors.material},
      -3px 1px 0 -1px ${contract.colors.borderLightest},
      -2px 0 0 -1px ${contract.colors.borderLightest},
      2px 1px 0 -1px ${contract.colors.borderDark},
      2px 0 0 -1px ${contract.colors.borderDarkest},
      3px 1px 0 -1px ${contract.colors.borderDarkest}`,
    selectors: {
      '&:first-child': {
        marginLeft: contract.space[4],
      },
    },
  },
  variants: {
    active: {
      true: {
        marginTop: calc.negate(contract.space[2]),
        marginLeft: calc.negate(contract.space[2]),
        marginBottom: contract.space[1],
        '&:first-child': { marginLeft: contract.space[2] },
        boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
        -2px 1px 0 -1px ${contract.colors.material},
        -0.5px 3px 0 0.5px ${contract.colors.material},
        -2px 2px 0 0 ${contract.colors.borderLightest},
        -2px 0 0 -1px ${contract.colors.borderLightest},
        1px 2px 0 0 ${contract.colors.borderDark},
        2px 0 0 -1px ${contract.colors.borderDarkest},
        2px 2px 0 0 ${contract.colors.borderDarkest}`,
      },
    },
  },
});

globalStyle(`${tab.classNames.variants.active.true} + li`, {
  marginLeft: 0,
  boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
    2px 1px 0 -1px ${contract.colors.borderDark},
    2px 0 0 -1px ${contract.colors.borderDarkest},
    3px 1px 0 -1px ${contract.colors.borderDarkest}`,
});
