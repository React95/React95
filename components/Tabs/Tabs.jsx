import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Tab from './Tab';

const Navbar = styled.ol`
  display: flex;
  padding: 0;
  margin: 0;
`;

const NavContainer = styled.div`
  padding: 12px;
  margin-left: 1px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff, 2px 1px 0 -1px #868a8e,
    1px 1px 0 0px #868a8e, 0px 1px 0 0px #868a8e, 2px 0px 0 -1px #000,
    2px 2px 0 0px #000, 0px 2px 0 0px #000, 0px 1px 0px 1px #ffffff,
    2px 0px 0 0px #000;
`;

const If = ({ condition, children }) => condition && children;

const Tabs = ({ children, style }) => {
  const firstTab = React.Children.toArray(children)[0];
  const [activeTab, setActiveTab] = useState(firstTab.props.title);

  return (
    <React.Fragment>
      <Navbar style={style}>
        {React.Children.map(children, (child) => {
          const { title, disabled } = child.props;

          return (
            <Tab
              key={title}
              title={title}
              activeTab={activeTab}
              onClick={!disabled && (() => setActiveTab(title))}
            />
          );
        })}
      </Navbar>

      <NavContainer style={style}>
        {React.Children.map(children, child => (
          <If condition={child.props.title === activeTab}>
            {child.props.children}
          </If>
        ))}
      </NavContainer>
    </React.Fragment>
  );
};

Tabs.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

export default Tabs;
