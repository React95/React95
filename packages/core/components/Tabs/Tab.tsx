import React, { forwardRef } from 'react';

import { tab } from './Tabs.css';
import { Frame, FrameProps } from '../Frame/Frame';

export type TabProps = {
  activeTab?: string;
  title: string;
  disabled?: boolean;
  onClick?(e: React.MouseEvent): void;
} & React.HTMLProps<HTMLLIElement> &
  Omit<FrameProps<'li'>, 'as'>;

export const Tab = forwardRef<HTMLLIElement, TabProps>(
  ({ activeTab, title, ...rest }, ref) => (
    <Frame
      {...rest}
      className={tab({ active: activeTab === title })}
      ref={ref}
      as="li"
    >
      {title}
    </Frame>
  ),
);
