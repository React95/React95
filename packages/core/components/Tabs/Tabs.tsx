import React, { forwardRef } from 'react';
import cn from 'classnames';

import Frame from '../Frame';
import { Tab, TabProps } from './Tab';
import { navContainer, navbar } from './Tabs.css';
import { FrameProps } from '../Frame/Frame';

export type TabsProps = {
  defaultActiveTab?: string;
  children: React.ReactElement<TabProps> | Array<React.ReactElement<TabProps>>;

  onChange?(title: string, e: React.MouseEvent): void;
} & Omit<FrameProps<'ol'>, 'as'>;

export const Tabs = forwardRef<HTMLOListElement, TabsProps>(
  ({ children, defaultActiveTab, onChange, ...rest }, ref) => {
    const [firstTab] = React.Children.toArray(children) as Array<
      React.ReactElement<TabProps>
    >;
    const [activeTab, setActiveTab] = React.useState(
      defaultActiveTab || firstTab.props.title,
    );

    return (
      <>
        <Frame
          {...rest}
          className={cn(navbar, rest.className)}
          as="ol"
          ref={ref}
        >
          {React.Children.map(
            children,
            (child: React.ReactElement<TabProps>) => {
              const { title, disabled } = child.props;

              return (
                <Tab
                  key={title}
                  {...child.props}
                  activeTab={activeTab}
                  onClick={e => {
                    if (!disabled) {
                      if (onChange) {
                        onChange(title, e);
                      }
                      setActiveTab(title);
                    }
                  }}
                />
              );
            },
          )}
        </Frame>

        <Frame className={navContainer} width={rest.width || rest.w}>
          {React.Children.map(
            children,
            (child: React.ReactElement<TabProps>) =>
              child.props.title === activeTab && child.props.children,
          )}
        </Frame>
      </>
    );
  },
);
