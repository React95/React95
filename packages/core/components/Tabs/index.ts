import { Tab } from './Tab';
import { Tabs } from './Tabs';

export { Tab, Tabs };
