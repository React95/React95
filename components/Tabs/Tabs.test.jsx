import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import Tab from './Tabs';
import Tabs from './Tabs';

describe('<Tabs />', () => {
  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <Tabs>
          <Tab title="example">
            <p>Example text</p>
          </Tab>
        </Tabs>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Active Tab', () => {
    const TabComponent = () =>
      render(
        <Tabs>
          <Tab title="first">
            <p>first text</p>
          </Tab>
          <Tab title="second">
            <p>second text</p>
          </Tab>
        </Tabs>
      );

    it('should have first Tab active', () => {
      const { queryByText } = TabComponent();
      expect(queryByText('first text')).toBeInTheDocument();
    });

    it('should active second Tab when Tab is clicked', () => {
      const { queryByText } = TabComponent();
      expect(queryByText('first text')).toBeInTheDocument();

      fireEvent.click(queryByText('second'));

      expect(queryByText('first text')).not.toBeInTheDocument();
      expect(queryByText('second text')).toBeInTheDocument();
    });
  });
});
