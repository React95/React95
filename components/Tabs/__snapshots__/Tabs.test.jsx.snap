// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Snapshot should match snapshot 1`] = `
.c1 {
  list-style: none;
  padding: 3px 6px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff,-2px 1px 0 -1px #c3c7cb,-3px 1px 0px -1px #ffffff,-2px 0px 0px -1px #ffffff,2px 1px 0 -1px #868a8e,2px 0px 0 -1px #000,3px 1px 0 -1px #000;
  margin-right: 2px;
  margin-left: 2px;
  margin-top: -2px;
  margin-left: -2px;
  margin-bottom: 1px;
  box-shadow: inset 0px 1px 0px 0px #ffffff,-2px 1px 0 -1px #c3c7cb,-0.5px 3px 0 0.5px #c3c7cb,-2px 2px 0px 0px #fff,-2px 0px 0px -1px #ffffff,1px 2px 0 0px #868a8e,2px 0px 0 -1px #000,2px 2px 0 0px #000;
}

.c1:first-child {
  margin-left: 4px;
}

.c1:first-child {
  margin-left: 2px;
}

.c1 + li {
  margin-left: 0;
  box-shadow: inset 0px 1px 0px 0px #ffffff,2px 1px 0 -1px #868a8e,2px 0px 0 -1px #000,3px 1px 0 -1px #000;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  margin: 0;
}

.c2 {
  padding: 12px;
  margin-left: 1px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff,2px 1px 0 -1px #868a8e,1px 1px 0 0px #868a8e,0px 1px 0 0px #868a8e,2px 0px 0 -1px #000,2px 2px 0 0px #000,0px 2px 0 0px #000,0px 1px 0px 1px #ffffff,2px 0px 0 0px #000;
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1"
    >
      example
    </li>
  </ol>
  <div
    class="c2"
  >
    <p>
      Example text
    </p>
  </div>
</div>
`;
